qiskit/primitives/base_sampler.py

```diff
@@ -130,6 +138,30 @@ def __init__(
                     f"and circuits ({len(self._circuits)}"
                 )
 
+    def __init_subclass__(cls):
+        original_init_method = cls.__init__
+
+        @wraps(cls.__init__)
+        def init_wrapper(
```

Comment:
Do I understand correctly that the job of this is to make sure the circuits are wrapped into a list and to set the circuit ids? If yes, wouldn't it be easier to read to just have this in the initializers of the subclasses? It would be a bit of duplicated code but it might be more straightforward to understand and maintain 🙂

Reply:
That is correct, but is there a way to force a third party subclass initializer? This is the only way I could think of to force it from the base class, but if you have a better way, please let me know.

Reply:
I found the way. We can do this by passing the initializer of the base class take a list of `QuantumCircuit.name` and a list of observable's ids.

```python
    def __init__(
        self,
        circuits: Iterable[QuantumCircuit],
        parameters: Iterable[Iterable[Parameter]] | None = None,
    ):
```
->
```python
    def __init__(
        self,
        circuits: Iterable[QuantumCircuit],
        circuit_names: Sequence[str],
        parameters: Iterable[Iterable[Parameter]] | None = None,
    ):
```

Reply:
Is this your suggestion?

Reply:
I guess instead of "forcing" it, you could just document that derived classes should call `super().__init__(...)` from their own `__init__` methods.

Reply:
Yeah either what Blake suggested or just have a getter for the `circuit_name` that just gets them from the circuit attributes like
```python
@property
def circuit_names(self):
      if self._circuit_names is None:
          self._circuit_names = [....]
      return self.circuit_names
```